File: algorithms/Javascript/selectionSort.js
// O(n^2) Time | O(1) Space
function selectionSort(array) {
	for (let startIdx = 0; startIdx < array.length - 1; startIdx++) {
    let smallestIdx = startIdx;
    for (let i = startIdx + 1; i < array.length; i++) {
      if (array[smallestIdx] > array[i]) smallestIdx = i;
    }
    const temp = array[smallestIdx];
    array[smallestIdx] = array[startIdx];
    array[startIdx] = temp;
  }
  return array;
}

// starts with unsorted array and two sublists, sorted and unsorted. 
// iterate through unsorted list, finding smallest number
// append to the start of the sorted sublist
// rinse and repeat until the unsorted list is empty
// to do in place, simply swap smallest number to the startIdx on each iteration

File: algorithms/Python/selectionSort.py
# O(n^2) Time | O(1) Space
def selectionSort(array):
    idx = 0 # idx of first num (starting pos) of unordered sublist
    while idx < len(array) - 1: # when we get to end of array, it's finished
        smallestIdx = idx
        for i in range(idx + 1, len(array)): # iterate through unordered list
            if array[smallestIdx] > array[i]: # at each point, if num is greater than current num
                smallestIdx = i # then update the smallestidx to be the current idx
        array[idx], array[smallestIdx] = array[smallestIdx], array[idx]
        idx += 1
    return array